### ISSUE TITLE
nasp export --type fasta might be a bit slow.

### ISSUE THREAD

**Issue by KasperThystrup** [TGenNorth/NASP#23]
Thank you very much for the impressive and comprehensive work of this tool! Absolutely brilliant to have a tool that actually applies itself to the local job management system, without the requirement for the users to have extra thoughts on this matter! Bravo! :-)

Inspired by #21, I attempt to generate an alignment of all files which can be used as input for gubbins, and for that I am exporting the master.tsv matrix, as it contains all positions.

I admit that I have a lot of samples in the master.tsv matrix (it amounts to 13Gb), however it seems surprising that the export --type fasta function manages to generate ~500Mb of data within 30 minutes on my local server node.

The help text for nasp export doesn't provide information on whether multithreading is an option (or recommended). Is this something that needs to be added, or might there be another bottleneck in the export code?

Kind regards
K

**Comment by RaSieb** [comment]
Is there a difference between `nasp export` and `matrix_to_fasta.py`?
Because the python script seems to be quite a bit faster while I could not see any difference in the output except for the wrapping of sequences:
```
>time nasp export --type fasta matrices/master_masked.tsv > matrices/master_masked.tmp.fasta                                                                                                                                         
real    14m30.988s
user    5m29.987s
sys     9m0.700s

>time matrix_to_fasta.py -m matrices/master_masked.tsv -p master_masked.tmp2
Total SNPs: 2793003
number of SNPs after filtering: 1755570
number of parsimony-informative SNPs: 6931

real    7m36.314s
user    7m10.083s
sys     0m9.023s
```

**Comment by jasonsahl** [comment]
Export also creates VCF files. The matrix to fasta command produces a separate parsimony-informative matrix and associated FASTA, which could be useful for some applications. So a lot of overlap between methods, but some differences that could be helpful depending on your application

**Comment by RaSieb** [comment]
Thank you for your explanation @jasonsahl. That makes sense!
Is there a way to get from the master.tsv to a master_masked.fasta file? So masking and converting to fasta?

**Comment by jasonsahl** [comment]
No, not currently. There is certainly enough information in the master.tsv file to facilitate that, but it's not currently implemented. The best practice for this would be to create the master_masked.tsv and then convert to FASTA from there.

**Comment by RaSieb** [comment]
Thank you so much for your responses! 
I have one more related question. I can see that there are more polymorphic sites maintained in the master_masked than in the bestsnp. So from the master_masked, is there a way to directly extract the bestsnp, or in other words, how are the two related to each other?

**Comment by jasonsahl** [comment]
The master_masked has information for every site in the reference, whether polymorphic or monomorphic across all samples. If you remove all monomorphic sites, as well any sites found in duplicated regions, you get the missingdata.tsv matrix. If you then remove any site from the missingdata matrix that has a non nucleotide character, you get the bestsnp.tsv matrix. So the bestsnp matrix is a subset of the master_masked; you could likely filter the master_masked matrix, based on this information, but would require some additional scripting
